pkg/tfsandbox/runtime.go

```diff
@@ -0,0 +1,340 @@
+		reattach = &auxServer.ReattachInfo
+	}
+
+	// TODO[pulumi/pulumi-terraform-module#199] concurrent access to the plugin cache
+	// if err := setupPluginCache(tf); err != nil {
+	// 	 return nil, fmt.Errorf("error setting up plugin cache: %w", err)
+	// }
+
+	return &ModuleRuntime{
+		tf:          tf,
+		reattach:    reattach,
+		description: "Terraform CLI",
+		executable:  execPath,
+	}, nil
+}
+
+func fileExists(filename string) bool {
+	_, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return err == nil
+}
+
+func NewRuntimeFromExecutable(
+	ctx context.Context,
+	logger Logger,
+	workdir Workdir,
+	auxServer *auxprovider.Server,
+	moduleExecutor string) (*ModuleRuntime, error) {
+
+	workDir, err := workdirGetOrCreate(ctx, logger, workdir)
+	if err != nil {
+		return nil, err
+	}
+	tf, err := tfexec.NewTerraform(workDir, moduleExecutor)
+	if err != nil {
+		return nil, fmt.Errorf("error creating a tofu executor: %w", err)
+	}
+
+	var reattach *tfexec.ReattachInfo
+	if auxServer != nil {
+		reattach = &auxServer.ReattachInfo
+	}
+
+	return &ModuleRuntime{
+		tf:          tf,
+		reattach:    reattach,
+		executable:  moduleExecutor,
+		description: fmt.Sprintf("module runtime from executable %s", moduleExecutor),
+	}, nil
+}
+
+// PickModuleRuntime will return a ModuleRuntime based on the provided moduleExecutor.
+// if executor: <path-to-executable>, it will create a runtime from that executable.
+// if executor: opentofu[@version] || tofu[@version], it will create a tofu runtime.
+// where version is optional, if not provided it will use the latest version of tofu.
+// anything else will default to a terraform runtime.
+func PickModuleRuntime(
+	ctx context.Context,
+	logger Logger,
+	workdir Workdir,
+	auxServer *auxprovider.Server,
+	moduleExecutor string) (*ModuleRuntime, error) {
+
+	// check if the module executor is a path to an existing executable
+	if fileExists(moduleExecutor) {
+		logger.Log(ctx, Debug, fmt.Sprintf("Using module executor from path: %s", moduleExecutor))
+		return NewRuntimeFromExecutable(ctx, logger, workdir, auxServer, moduleExecutor)
+	}
+
+	if strings.HasPrefix(moduleExecutor, "opentofu") || strings.HasPrefix(moduleExecutor, "tofu") {
+		resolveOptions := tofuresolver.ResolveOpts{}
+		var version string
+		if parts := strings.Split(moduleExecutor, "@"); len(parts) == 2 {
+			// If the module executor is in the format "opentofu@version" or "tofu@version",
+			// we extract the version and set it in the resolve options.
+			version = parts[1]
+			parsedVersion, err := semver.Parse(version)
+			if err != nil {
+				return nil, fmt.Errorf("error parsing version %q for %s: %w", parts[1], parts[0], err)
+			}
+			resolveOptions.Version = &parsedVersion
+		}
+
+		logger.Log(ctx, Debug, fmt.Sprintf("Using module executor %s %s", moduleExecutor, version))
+		return NewTofu(ctx, logger, workdir, auxServer, resolveOptions)
+	}
+
+	// check if the module executor is a path to an existing executable
```

Comment:
This looks like a TODO that has not been implemented yet, right? It's only looking for terraform.exe, if the user said "executor": "my-terraform" this would not be looking for "my-terraform.exe". We can punt it for later just comment hygiene around "check if the module executor is a path".